### DISCUSSION TITLE
Windows service and updating from old OpenDCS

### DISCUSSION THREAD

**Discussion by gamiller0-tva** [opendcs/opendcs#629] [Q&A]
Quick question.  Is the JSW Service wrapper still used for LRGS and Routing on windows?  I see references to it in the docs, but not seeing it in the most recent install.  

Thanks,
-Gabe

**Comment by MikeNeilson** [comment]
We haven't been maintaining anything that's compiled code recently.

If you need to run the OpenDCS software as a service in windows external tools like https://commons.apache.org/proper/commons-daemon/ are probably going to work better than anything we'd be able to try and maintain here.

That said we're open to a solution offered, would just need sufficient tests worked into the automated workflow to ensure maintainability.

**Comment by gamiller0-tva** [comment]
Thanks Mike.  I'll do some exploring.  Do you know what other windows users are doing currently?  Or are most running on linux with unit files?

**Reply by MikeNeilson** [reply]
For services, as far as a I now it's mostly Linux, or other unix. And even there it's mostly `nohup <command> &`. I am using a systemd unit file on one of my personal projects, but now that you ask I don't think it's the one that's in the repo, I'll have to check later and update if necessary.

**Comment by gamiller0-tva** [comment]
An example systemd unit file would be helpful to see, should inform how to setup the service for the windows side as well.  I haven't worked with commons-daemon before, but hopefully its easy to setup.

**Reply by MikeNeilson** [reply]
I haven't heard bad things about it. But haven't used it myself. 

Oh, not sure if it would work in your environment or not. We do also have docker images. They're hosted on the GitHub container registry. I've only tested them in a linux docker instance, but I don't see why they wouldn't work in a windows hosted instance or WSL.

https://github.com/orgs/opendcs/packages

A single box were you can create a volume for lrgs_home things should work just fine. I haven't gotten it to work in the AWS environment I have access to with that block storage, but we do have plans to extend the archive storage to both SQL database and S3 compatible storage.

**Reply by gamiller0-tva** [reply]
Thanks for your quick responses and answering all my questions.  Do you know if there are existing start/stop classes that could be used for commons-daemon/procrun?  Looks like that is required.  

i.e.

class MyClass;
// N.B. error handling not shown
static void main(String [] args){
    String mode = args[0];
    if ("start".equals(mode){
        // process service start function
    }
    etc.
}

**Reply by MikeNeilson** [reply]
Unfortunately I wasn't aware that was required, the good news is it's probably not going to be difficult to create, the bad news is I'm not sure when I or anyone else currently working on code would have time.

Is this something you might be inclined to help with directly?

**Reply by gamiller0-tva** [reply]
Possibly.  I'll see if anyone on our team knows java and can do it.
